### PR TITLE
[RFC] Demux tokens

### PR DESCRIPTION
**Pull request by juimonen**
In these 2 commits I'm trying to show/demo how you can parse more complex tuple arrays, example being mux/demux.

### CODE REVIEW

**Comment by juimonen** [comment]
fw counterpart: https://github.com/thesofproject/sof/pull/2345

**Comment by ranj063** [comment]
@juimonen Could you please paste the conf file for an example tplg that we're trying to parse here?I am having a hard time understanding the problem without it

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -2092,6 +2171,30 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+	case SOF_COMP_EQ_FIR:
+		/* fallthrough */
+	case SOF_COMP_EQ_IIR:
+		/* fallthrough */
+	case SOF_COMP_KEYWORD_DETECT:
+		/* fallthrough */
+	case SOF_COMP_KPB:
+		/* fallthrough */
+	case SOF_COMP_SELECTOR:
+		break;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: why would we want to list out everything here? this doesn't scale. Just test for mux/demux and break in the default case (all others).

File: sound/soc/sof/topology.c [hunk 2]
```diff
@@ -2092,6 +2171,30 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 
 	sof_dbg_comp_config(scomp, &process->config);
 
+	/* possible tuplet and data parsing for process components */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: note on comment half a page above "configure iir IPC message" is probably just wrong/obsolete?

File: sound/soc/sof/topology.c [hunk 3]
```diff
@@ -2017,6 +2042,60 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
...
+				       varray,
+				       le32_to_cpu(private->size) - pos);
+		if (ret < 0) {
+			dev_err(scomp->dev, "error: mux stream tokens %d\n",
+				private->size);
+			return ret;
+		}
+
+		/* track position for multiple same type arrays */
+		pos += ret;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: it's really horrible to mix position and error status, sorry.

e.g. here ret could be zero, so the position is NOT updated...

Don't do it, it's just error prone and difficult to maintain.

File: sound/soc/sof/topology.c [hunk 4]
```diff
@@ -2017,6 +2042,60 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
...
+				       ARRAY_SIZE(mux_stream_tokens),
+				       varray,
+				       le32_to_cpu(private->size) - pos);
+		if (ret < 0) {
+			dev_err(scomp->dev, "error: mux stream tokens %d\n",
+				private->size);
+			return ret;
+		}
+
+		/* track position for multiple same type arrays */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: I have idea what 'multiple same type arrays' are? 

As opposed to single same type array? or multiple different type arrays?

File: sound/soc/sof/topology.c [hunk 5]
```diff
@@ -2017,6 +2042,60 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
...
+			       ARRAY_SIZE(mux_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret < 0) {
+		dev_err(scomp->dev, "error: parse mux tokens failed %d\n",
+			private->size);
+		return ret;
+	}
+
+	/* parse mux stream config arrays */
+	for (i = 0; i < mc->num_streams; i++) {
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: if there a range check on mc->streams? Or could a bad topology break things here?

File: include/sound/sof/topology.h [hunk 6]
```diff
@@ -174,6 +174,22 @@ struct sof_ipc_comp_asrc {
 	uint32_t reserved[4];
 } __attribute__((packed));
 
+struct mux_stream_data {
+	uint32_t data_id;
+	uint32_t pipeline_id;
+	uint8_t num_channels;
+	uint8_t mask[8];
+	uint8_t reserved[3]; /* padding */
+} __packed;
```
Comment (plbossart) [inline include/sound/sof/topology.h:183]: I don't think we have u8 in IPC definitions? Are you sure we're not going to have endianness issues or other shenanigangs?

File: include/sound/sof/topology.h [hunk 7]
```diff
@@ -174,6 +174,22 @@ struct sof_ipc_comp_asrc {
 	uint32_t reserved[4];
 } __attribute__((packed));
 
+struct mux_stream_data {
+	uint32_t data_id;
+	uint32_t pipeline_id;
+	uint8_t num_channels;
+	uint8_t mask[8];
```
Comment (plbossart) [inline include/sound/sof/topology.h:181]: what does the mask represent?

File: sound/soc/sof/topology.c [hunk 8]
```diff
@@ -936,6 +953,11 @@ static int sof_parse_tokens(struct snd_soc_component *scomp,
 		/* next array */
 		array = (struct snd_soc_tplg_vendor_array *)((u8 *)array
 			+ asize);
+		pos += asize;
+
+		/* return position where the array was found */
+		if (found == count)
+			return pos;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: doesnt' look like any of the routines below actually use the position for anything, they just test the status. Why is this needed? or what I am missing?

File: sound/soc/sof/topology.c [hunk 9]
```diff
@@ -751,14 +751,15 @@ static const struct sof_topology_token led_tokens[] = {
 	 get_token_u32, offsetof(struct snd_sof_led_control, direction), 0},
 };
 
-static void sof_parse_uuid_tokens(struct snd_soc_component *scomp,
-				  void *object,
-				  const struct sof_topology_token *tokens,
-				  int count,
-				  struct snd_soc_tplg_vendor_array *array)
+static int sof_parse_uuid_tokens(struct snd_soc_component *scomp,
```
Comment (plbossart) [inline sound/soc/sof/topology.c:779]: comment on commit message: I have absolutely no idea if you describe a data corruption issue ("copied on top"), a bad list parsing (returns last instead of first).

What are 'multiple similar tuple arrays'? why is this useful?
What are you trying to fix or enable really?

File: sound/soc/sof/topology.c (hunk 3 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok, yes this is really good comment, I need to pass the position as parameter.

File: sound/soc/sof/topology.c (hunk 4 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: If you have 2 similar tuples arrays in one component's topology, current parser will always parse the last one. So it will parse the 1st one, but continue and parse the 2nd on top of the 1st. Mux is an example of this kind of component, because it may contain multiple stream config arrays. 

I'm not sure if we currently any other component with this kind of structure, but we may in the future. Also I'm not aware any other way how this should be handled in alsa topology, is it forbidden to define multiple similar tuples arrays for one component? At least currently the alsa tools are fine with it.

File: include/sound/sof/topology.h (hunk 6 above)
Comment (juimonen) [reply include/sound/sof/topology.h:183]: no I'm not. Actually the 8 and 3 in the mask and reserved are also dubious as we have variable size for those in FW depending on platform (for e.g. 8 for apl, 4 for byt). So actually I would like us also to rationalize the mux/demux structs, if they are made public in kernel, so that it would be same for all platforms and have some sane types. I mean pipeline_id is uint32 and channels is uint8? Are we going to have 2 million pipelines? So why don't just set everything to uint16 and fix the max channels to 8?

I think the space save we get from these hacks is minimal. So this is more for mux/demux component developer, but as these structs could be mirrored in kernel, they should take into account other issues as minimal size also...

File: include/sound/sof/topology.h (hunk 7 above)
Comment (juimonen) [reply include/sound/sof/topology.h:181]: so these are copied from FW and should be explained there. As I understand it, it is a channel mask for this stream, so which stream and channel is copied and where? @akloniex any comments?

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: the mux/demux parsing uses it. nobody else doesn't currently.

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:779]: Yeah, sorry for being too sparse. I tried to explain this above also... so I don't think currently the parser handles multiple similar tuples arrays following each other in topology. Every array is parsed but copied on top of the previous one. And you can't do that easily as parser doesn't tell that it has found something or the position of the found array. So our parser will always run to the end of given data section despite it found 1,2,3,4 instances of the array and the last one of those is what you get. AFAIK currently there is just no way of parsing the array number 1,2,3 as the parser will run to the end of data.

As said we've not had this kind of situation, because our components don't currently have this kind of multiple array structure. Mux/demux is the first one with this kind of more complex structure.

**Comment by juimonen** [comment]
@ranj063 check the demux part in following conf file...
[sof-apl-demux-pcm512x.txt](https://github.com/thesofproject/linux/files/4142277/sof-apl-demux-pcm512x.txt)

File: sound/soc/sof/topology.c [hunk 10]
```diff
@@ -773,22 +774,27 @@ static void sof_parse_uuid_tokens(struct snd_soc_component *scomp,
 			/* match token id */
 			if (tokens[j].token != le32_to_cpu(elem->token))
 				continue;
+			else
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: no need for an "else" - the below code is also in the same "else" case (implicitly). Your `found++;` is just one of the statements, that will be executed, if the branch isn't taken.

File: sound/soc/sof/topology.c [hunk 11]
```diff
@@ -803,26 +809,31 @@ static void sof_parse_string_tokens(struct snd_soc_component *scomp,
...
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: ditto

File: sound/soc/sof/topology.c [hunk 12]
```diff
@@ -840,6 +851,8 @@ static void sof_parse_word_tokens(struct snd_soc_component *scomp,
...
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: here too

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (lyakh) [inline sound/soc/sof/topology.c]: nitpick: you could add msd++ to the loop statement:
```
for (i = 0; i < mc->num_streams; i++, msd++) {
```
to make it more obvious, that each loop iteration advances `msd`

File: sound/soc/sof/topology.c [hunk 13]
```diff
@@ -2017,6 +2042,60 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
...
+		}
+
+		/* track position for multiple same type arrays */
+		pos += ret;
+		msd++;
+	}
+
+	/* get the channel byte masks after stream arrays */
+	msd -= mc->num_streams;
+	for (i = 0; i < mc->num_streams; i++) {
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: ditto

File: sound/soc/sof/topology.c (hunk 1 above)
Comment (lyakh) [reply sound/soc/sof/topology.c]: ...implicit default, since `type` isn't an enum, no explicit `default` case is really needed.

File: sound/soc/sof/topology.c [hunk 14]
```diff
@@ -2092,6 +2171,30 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+	case SOF_COMP_DEMUX:
+		offset = sof_process_load_mux(scomp, process, private);
+		break;
+	}
+
+	if (offset < 0) {
+		dev_err(scomp->dev, "error: parse process tokens failed %d\n",
+			le32_to_cpu(private->size));
+		goto err;
+	}
```
Comment (lyakh) [inline sound/soc/sof/topology.c:2198]: better move this inside the `switch` - only there `offset` can be negative - right after the assignment.

File: sound/soc/sof/topology.c (hunk 1 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok, removed listing of all existing components and only mux/demux remains.

File: sound/soc/sof/topology.c (hunk 2 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: yes some oldie, has to be fixed in separate pr was already before this PR...

File: sound/soc/sof/topology.c (hunk 3 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: fixed now. There's read_bytes argument given to parse_tokens now and return value is just what it was before.

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: AFAIK this is parsed against total data size and tuples array sizes. So we should not go beyond the data area without error...

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 11 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 12 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok, actually added even some more, hope it doesn't get confusing :)

File: sound/soc/sof/topology.c (hunk 13 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok, so the added things we're here, not above.

File: sound/soc/sof/topology.c (hunk 14 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:2198]: ok

**Comment by juimonen** [comment]
@plbossart @ranj063 @lyakh updated. Actually I want to achieve the following things:
1) sof_parse_tokens should return when it has found all tokens (found == count)
2) sof_parse_tokens should return the current parse position
3) I'm using the above to parse multiple similar tuples arrays in mux/demux

final observation: I don't get why we are checking (ret != 0) everywhere with sof_parse_tokens? I don't see it could return positive values..

**Comment by ranj063** [comment]
@juimonen I am still trying to understand what the actual problem is. So please bear with my stupid questions here.

Looking at the conf file you sent, you have the below. Is your problem that you have 2 stream data tuples that have similar sets of tokens? ex: mux_stream_data_1 and mux_stream_data_2?

If so, this is a problem we have already solved in the case of DMIC. We have 2 sets of PDM config tokens and we parse both sets depending on the number of PDM's in the DMIC config. Could you please take a look?
```
SectionData."mask_priv_1" {
	       bytes "0x01,0x04,0x08,0x10,0x18,0x20,0x40,0x80"
}
SectionVendorTuples."mux_stream_data_1_tuples_w" {
	tokens "sof_mux_tokens"
	tuples."word" {
		SOF_TKN_MUX_STREAM_PIPELINE_ID 	"1"
		SOF_TKN_MUX_STREAM_CHANNELS		"2"
	}
}
SectionData."mux_stream_data_1" {
	tuples "mux_stream_data_1_tuples_w"
}

SectionData."mask_priv_2" {
	       bytes "0x01,0x04,0x08,0x10,0x18,0x20,0x40,0x80"
}
SectionVendorTuples."mux_stream_data_2_tuples_w" {
	tokens "sof_mux_tokens"
	tuples."word" {
		SOF_TKN_MUX_STREAM_PIPELINE_ID 	"5"
		SOF_TKN_MUX_STREAM_CHANNELS		"2"
	}
}
SectionData."mux_stream_data_2" {
	tuples "mux_stream_data_2_tuples_w"
}
```

File: sound/soc/sof/topology.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/topology.c]: can you give an example of what 'similar tuple arrays in one components's topology' mean?

These are really vague concepts for me, and I am not sure why this is exposed by the mux/demux support?

Comment (ranj063) [reply sound/soc/sof/topology.c]: @plbossart @juimonen If I understand it correctly, I think the problem is similar to what we've done for the PDM controller config in the topology for DMIC config already. We're looking at something like below where we have multiple sets of tokens.

https://github.com/thesofproject/sof/pull/2345#issuecomment-581502285

File: sound/soc/sof/topology.c [hunk 15]
```diff
@@ -936,6 +952,13 @@ static int sof_parse_tokens(struct snd_soc_component *scomp,
 		/* next array */
 		array = (struct snd_soc_tplg_vendor_array *)((u8 *)array
 			+ asize);
+
+		/* update parse position */
+		*read_bytes = *read_bytes + asize;
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: *read_bytes += asize;

**Comment by juimonen** [comment]
@ranj063 yes I saw the dmic stuff in sof_parse_tokens. I think my solution is more generic, but that's just my personal opinion... actually I would like to convert the dmic part similar to this solution.

File: sound/soc/sof/topology.c [hunk 16]
```diff
@@ -936,6 +952,13 @@ static int sof_parse_tokens(struct snd_soc_component *scomp,
...
+
+		/* return when all tokens have been found */
+		if (found == count)
+			return 0;
```
Comment (lyakh) [inline sound/soc/sof/topology.c:986]: you could just add this to your loop condition as `while (priv_size > 0 && found != count)`

File: sound/soc/sof/topology.c [hunk 17]
```diff
@@ -962,6 +985,7 @@ static int sof_control_load_volume(struct snd_soc_component *scomp,
 		container_of(hdr, struct snd_soc_tplg_mixer_control, hdr);
 	struct sof_ipc_ctrl_data *cdata;
 	int tlv[TLV_ITEMS];
+	int read_bytes = 0;
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: you initialise `*read_bytes` inside `sof_parse_tokens()`, so, no need to also initialise it in each caller function

File: sound/soc/sof/topology.c [hunk 18]
```diff
@@ -2029,6 +2054,61 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
...
+static int sof_process_load_mux(struct snd_soc_component *scomp,
+				struct sof_ipc_comp_process *process,
+				struct snd_soc_tplg_private *private,
+				int *bytes_written)
+{
+	struct sof_mux_config *mc = (struct sof_mux_config *)&process->data;
+	struct mux_stream_data *msd = (struct mux_stream_data *)
+		((u8 *)process->data + sizeof(struct sof_mux_config));
+	struct snd_soc_tplg_vendor_array *varray;
+	int read_bytes = 0;
```
Comment (lyakh) [inline sound/soc/sof/topology.c:2069]: ditto in this patch

File: sound/soc/sof/topology.c (hunk 15 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 16 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:986]: @lyakh I have problem with this function, so should we return error value if not all tokens are found? I think we shouldn't have situation where some tokens are defined but some are not? then I could return here 0 and something like -EINVAL at the end...

File: sound/soc/sof/topology.c (hunk 17 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 18 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:2069]: ok

File: sound/soc/sof/topology.c [hunk 19]
```diff
@@ -1704,15 +1735,15 @@ static int sof_widget_load_pga(struct snd_soc_component *scomp, int index,
...
+			       le32_to_cpu(private->size), &read_bytes);
 	if (ret != 0) {
 		dev_err(scomp->dev, "error: parse volume tokens failed %d\n",
 			private->size);
 		goto err;
 	}
 	ret = sof_parse_tokens(scomp, &volume->config, comp_tokens,
 			       ARRAY_SIZE(comp_tokens), private->array,
-			       le32_to_cpu(private->size));
+			       le32_to_cpu(private->size), &read_bytes);
```
Comment (ranj063) [inline sound/soc/sof/topology.c:1771]: @juimonen whats the purpose of read_bytes here? How is it used?

File: sound/soc/sof/topology.c [hunk 20]
```diff
@@ -2029,6 +2054,61 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
...
+
+	/* get the channel byte masks after stream arrays */
+	msd -= mc->num_streams;
+	for (i = 0; i < mc->num_streams; i++, msd++, pos += 8)
+		memcpy(&msd->mask[0], (u8 *)private->array + pos, 8);
+
+	*bytes_written = sizeof(struct sof_mux_config) +
+		mc->num_streams * sizeof(struct mux_stream_data);
+
+	return ret;;
```
Comment (ranj063) [inline sound/soc/sof/topology.c]: duplicate ;

File: sound/soc/sof/topology.c [hunk 21]
```diff
@@ -2029,6 +2054,61 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
...
+		dev_err(scomp->dev, "error: parse mux tokens failed %d\n",
+			private->size);
+		return ret;
+	}
+
+	/* parse mux stream config arrays */
+	for (i = 0; i < mc->num_streams; i++, msd++) {
+		varray = (struct snd_soc_tplg_vendor_array *)
+			((u8 *)private->array  + pos);
+		ret = sof_parse_tokens(scomp, msd, mux_stream_tokens,
```
Comment (ranj063) [inline sound/soc/sof/topology.c:2088]: @juimonen If I am not wrong, you will end up overwriting msd with every stream_data array tokens that are parsed here isnt it? Did you miss incrementing msd addr inside the loop?

File: sound/soc/sof/topology.c [hunk 22]
```diff
@@ -2029,6 +2054,61 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
...
+				private->size);
+			return ret;
+		}
+		pos += read_bytes;
+	}
+
+	/* get the channel byte masks after stream arrays */
+	msd -= mc->num_streams;
+	for (i = 0; i < mc->num_streams; i++, msd++, pos += 8)
+		memcpy(&msd->mask[0], (u8 *)private->array + pos, 8);
```
Comment (ranj063) [inline sound/soc/sof/topology.c:2104]: mask[0]? or should this be mask[i]?

File: sound/soc/sof/topology.c (hunk 21 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:2088]: oops sorry didnt realize it was in the for (). please ignore

File: sound/soc/sof/topology.c (hunk 22 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:2104]: oh this is very confusing. looks fine. sorry

File: sound/soc/sof/topology.c (hunk 19 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:1771]: so the read_bytes will give you the number of how many bytes from the beginning of the data area the last array was found and parsed. So you can parse next array from that point and not start from beginning again and again... You can see the example in the demux parsing in this pr.

File: sound/soc/sof/topology.c (hunk 20 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: oh, ok fixed now.

**Comment by juimonen** [comment]
@ranjani was checking again the dmic topology after talking to you.... So it is done little bit different than my implementation:

1) In dmic there's only 1 array with multiple tuplet sets inside it, I have here several arrays with 1 set of tuples inside.

2) dmic tuplet parse implementation assumes that the ipc struct is similar in contents of the struct coming from topology (so you can just increment 32bit slots and assume the value is inserted to correct place). As for example the demux "ipc" structs have all kinds of paddings etc. so you can't immediately assume that just incrementing the dst pointer we get correct placement.

So, we have quite a lot of assumptions in the parsing, not sure what would be the correct thing to do...

**Comment by juimonen** [comment]
@ranj063 @plbossart @lyakh thank you for you reviews. After talking with Ranjani I went for different approach: https://github.com/thesofproject/linux/pull/1766

It is changing the existing dmic "multi-item" parsing and doesn't touch the demux at all, as I also learned the demux interface is in flux. In the new PR I try to generalize the "multi-item" parsing so the demux (or others) should be fairly easy to add.

**Comment by juimonen** [comment]
so closing...
